pkg/spi-shared/config/userconfig.go

```diff
@@ -14,9 +14,124 @@
+
+	lg.V(logs.DebugLevel).Info("looking for sp configuration secrets", "namespace", tokenNamespace, "spname", spType.Name, "sphost", spHost)
+
+	secrets := &corev1.SecretList{}
+	if listErr := k8sClient.List(ctx, secrets, client.InNamespace(tokenNamespace), client.MatchingLabels{
+		v1beta1.ServiceProviderTypeLabel: string(spType.Name),
+	}); listErr != nil {
+		if kuberrors.IsForbidden(listErr) {
+			lg.Info("not enough permissions to list the secrets")
+			return false, nil, nil
+		} else if kuberrors.IsUnauthorized(listErr) {
+			lg.Info("request is not authorized to list the secrets in user's namespace")
+			return false, nil, nil
+		} else {
+			return false, nil, fmt.Errorf("failed to list oauth config secrets: %w", listErr)
+		}
+	}
+
+	lg.V(logs.DebugLevel).Info("found secrets with oauth configuration", "count", len(secrets.Items))
+
+	if len(secrets.Items) < 1 {
+		return false, nil, nil
+	}
+
+	var oauthSecretWithoutHost *corev1.Secret
+	var oauthSecretWithHost *corev1.Secret
+
+	// go through all found oauth secret configs
+	for _, oauthSecret := range secrets.Items {
+		// if we find one labeled for sp host, we take it
+		if labelSpHost, hasLabel := oauthSecret.ObjectMeta.Labels[v1beta1.ServiceProviderHostLabel]; hasLabel {
+			if labelSpHost == spHost {
+				if oauthSecretWithHost != nil { // if we found one before, return error because we can't tell which one to use
+					return false, nil, errMultipleMatchingSecrets
+				}
+				oauthSecretWithHost = oauthSecret.DeepCopy()
+			}
+		} else { // if we found one without host label we check if it matches with default sp host value, then we can save it for later
+			if spHost != spType.DefaultHost {
+				continue
+			}
+			if oauthSecretWithoutHost != nil { // if we found one before, return error because we can't tell which one to use
+				return false, nil, errMultipleMatchingSecrets
+			}
+			oauthSecretWithoutHost = oauthSecret.DeepCopy()
+		}
+	}
+
+	if oauthSecretWithHost != nil {
+		return true, oauthSecretWithHost, nil
+	} else if oauthSecretWithoutHost != nil {
+		return true, oauthSecretWithoutHost, nil
+	} else {
+		return false, nil, nil
+	}
+}
+
+func createServiceProviderConfigurationFromSecret(configSecret *corev1.Secret, baseUrl string, spType ServiceProviderType) *ServiceProviderConfiguration {
+	return &ServiceProviderConfiguration{
+		ServiceProviderType:    spType,
+		ServiceProviderBaseUrl: baseUrl,
+		Extra:                  map[string]string{},
+		OAuth2Config:           initializeOAuthConfigFromSecret(configSecret, spType),
+	}
+}
+
+func initializeOAuthConfigFromSecret(secret *corev1.Secret, spType ServiceProviderType) *oauth2.Config {
+	oauthCfg := &oauth2.Config{
+		Endpoint: spType.DefaultOAuthEndpoint,
+	}
+	if clientId, has := secret.Data[oauthCfgSecretFieldClientId]; has {
+		oauthCfg.ClientID = string(clientId)
+	} else {
+		// in case we don't have client id, we consider configuration to not have oauth
```

Comment:
> Could it be worth logging this? So that users know they created malformed configuration secret.

Maybe even putting an annotation on the secret itself, because the secret author might not have access to our logs.

Reply:
Such secret isn't malformed. This PR is about oauth configuration being optional, so Secret without oauth information is perfectly valid.

Reply:
only case that might worth to report error to user is, if only one of clientId clientSecret is set.

Reply:
I've created an issue, this is not trivial https://issues.redhat.com/browse/SVPI-304